modules/tinymce/src/core/main/ts/undo/UndoManagerTypes.ts

```diff
@@ -26,6 +26,14 @@ export interface CompleteUndoLevel extends BaseUndoLevel {
 export type NewUndoLevel = CompleteUndoLevel | FragmentedUndoLevel;
 export type UndoLevel = NewUndoLevel & { bookmark: Bookmark };
 
+export interface Extra {
+  (callback1: () => void, callback2: () => void): void;
+  (callback1: () => Promise<void>, callback2: () => void): Promise<void>;
+  (callback1: () => void, callback2: () => Promise<void>): Promise<void>;
+  // eslint-disable-next-line @typescript-eslint/unified-signatures
+  (callback1: () => Promise<void>, callback2: () => Promise<void>): Promise<void>;
```

Comment:
I was hoping to let this help typescript be smarter and to avoid the internal issues we had.

Reply:
This is done backwards just add a new type in UndoManagerTypes. This is the only type you need.

`export type ExtraCallback = () => void | Promise<void>;`

Then use that all over for example:
```
export interface UndoManager {
  data: UndoLevel[];
  typing: boolean;
  add: (level?: Partial<UndoLevel>, event?: EditorEvent<any>) => UndoLevel | null;
  dispatchChange: () => void;
  beforeChange: () => void;
  undo: () => UndoLevel | undefined;
  redo: () => UndoLevel | undefined;
  clear: () => void;
  reset: () => void;
  hasUndo: () => boolean;
  hasRedo: () => boolean;
  transact: (callback: () => void) => UndoLevel | null;
  ignore: (callback: () => void) => void;
  extra: (callback1: ExtraCallback, callback2: ExtraCallback) => void;
}
```

Or in the RTC adaptor:
```
interface RtcAdaptor {
  init: {
    bindEvents: () => void;
  };
  undoManager: {
    beforeChange: (locks: Locks, beforeBookmark: UndoBookmark) => void;
    add: (
      undoManager: UndoManager,
      index: Index,
      locks: Locks,
      beforeBookmark: UndoBookmark,
      level?: Partial<UndoLevel>,
      event?: Event
    ) => UndoLevel | null;
    undo: (undoManager: UndoManager, locks: Locks, index: Index) => UndoLevel | undefined;
    redo: (index: Index, data: UndoLevel[]) => UndoLevel | undefined;
    clear: (undoManager: UndoManager, index: Index) => void;
    reset: (undoManager: UndoManager) => void;
    hasUndo: (undoManager: UndoManager, index: Index) => boolean;
    hasRedo: (undoManager: UndoManager, index: Index) => boolean;
    transact: (undoManager: UndoManager, locks: Locks, callback: () => void) => UndoLevel | null;
    ignore: (locks: Locks, callback: () => void) => void;
    extra: (undoManager: UndoManager, index: Index, callback1: ExtraCallback, callback2: ExtraCallback) => void;
  };
  formatter: {
    match: Formatter['match'];
    matchAll: Formatter['matchAll'];
    matchNode: Formatter['matchNode'];
    canApply: Formatter['canApply'];
    closest: Formatter['closest'];
    apply: Formatter['apply'];
    remove: Formatter['remove'];
    toggle: Formatter['toggle'];
    formatChanged: (registeredFormatListeners: Cell<RegisteredFormats>, formats: string, callback: FormatChangeCallback, similar?: boolean, vars?: FormatVars) => UnbindFormatChanged;
  };
  editor: {
    getContent: (args: GetContentArgs) => Content;
    setContent: (content: Content, args: SetContentArgs) => SetContentResult;
    insertContent: (value: string, details: InsertContentDetails) => string;
    addVisual: (elm?: HTMLElement) => void;
  };
  selection: {
    getContent: (format: ContentFormat, args: Partial<GetSelectionContentArgs>) => Content;
  };
  autocompleter: {
    addDecoration: (range: Range) => void;
    removeDecoration: () => void;
  };
  raw: {
    getModel: () => Optional<any>;
  };
}
```

You don't need to move any code around, add any extra interfaces except this one type. Then you do all the complex magic in the Operations.extra where it should be.

Reply:
Right now I see you need all that type wrangling to return a promise as the response maybe it's better to just make two functions then one that is async and one that isn't instead of having one function being able to accept both and return a promise or a void depending on input. Do we need it to support partial async or could it all be just one async thing since if the first callback is a promise then the whole thing will be async anyway.

Reply:
We still wish to preserve the `extra: (callback1: () => void, callback2: () => void) => void`.
All the other versions should return a promise. Which is what all the typing is about. Though typescript is quite unhelpful with that.

Reply:
Two functions would still make it hard to allow partial async you want need the first extra to be sync then the second insert async for markdown right? So it would then still need three cases and we don't want the first one to be async for no reason since that might let the browser render the initial level since we yield to the event loop for a while. Since this is an API that just we need could be just add that single case making the second callback async. Then at least we only have two types to deal with? I can't think of a case where we need to do something async on the first callback. If we do we can deal with that problem later.

Reply:
Thinking of it a bit more I think the best way to approach this is to make a new `extraAsync` function since then the return type will only be different for this function and we don't need to touch the current `extra` function at all so that one doesn't become complex. Then this new `extraAsync` function would just be that simple type of `extraAsync: (callback1: () => void | Promise<void>, calllback2: void | Promise<void>): Promise<void>` then you wouldn't need any of the weird overloads so everything becomes simpler. We could discuss if we need both callbacks to be async or not. These `extra` functions are pretty weird and I think it probably just us that use them so adding another extraAsync wouldn't confuse users of the API much since they are likely not using `extra` anyway. I also like that having a extra function tells you that the entire thing becomes async when you call it it hints that it returns a promise and it's a common practise in node apps etc to add the `Async` suffix.

Reply:
Or we could simplify the types so that the callbacks are either both sync or both return promises. Two possibilities instead of four. An extra `return Promise.resolved()` is not too hard to deal with as an integrator, and will make this all a lot easier to follow.

Reply:
`Two functions would still make it hard to allow partial async you want need the first extra to be sync then the second insert async for markdown right?` Other way around. We need the first one to be async for the markdown plugin, not the second one.
The problem with the types that go `(callback1: () => void | Promise<void>, calllback2: void | Promise<void>): Promise<void>` is that that includes the case of `(callback1: () => void, calllback2: () => void): Promise<void>`, which we wanted to return `void` and not `Promise<void>`.
extraAsync would certainly solve the issue there. If we wish to change it to that I can do that.

Reply:
Ok then you can make then both async no optional async at all that simplifies things. The only reason to try to make the first one non async is to potentially avoid flicker. The flicker will if it occurs be a potential blocker for this change though if that is the case we might need to consider other options.